crates/ruff_linter/src/rules/flake8_pyi/rules/bytestring_usage.rs

```diff
@@ -0,0 +1,83 @@
+use ruff_diagnostics::{Diagnostic, FixAvailability, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{self as ast, Expr};
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for the uses of `typing.ByteString` and `collections.abc.ByteString`.
+///
+/// ## Why is this bad?
+/// `ByteString` has been deprecated since Python 3.9 and will be removed in
+/// Python 3.14. The Python documentation recommends using either
+/// `collections.abc.Buffer` (for Python >=3.12) or a union like
+/// `bytes | bytearray | memoryview` instead.
+///
+/// ## Example
+/// ```python
+/// from typing import ByteString
+/// ```
+///
+/// Use instead:
+/// ```python
+/// from collections.abc import Buffer
+/// ```
+///
+/// ## References
+/// - [Python documentation: The `ByteString` type](https://docs.python.org/3/library/typing.html#typing.ByteString)
+#[violation]
+pub struct ByteStringUsage {
+    full_name: String,
+}
```

Comment:
Rather than storing the full name as a string on the violation struct, I'd be tempted to do something like this, so that we only record the module where `ByteString` came from:

```rs
#[violation]
pub struct ByteStringUsage {
    origin: ByteStringOrigin,
}

impl Violation for ByteStringUsage {
    const FIX_AVAILABILITY: FixAvailability = FixAvailability::None;

    #[derive_message_formats]
    fn message(&self) -> String {
        let ByteStringUsage { origin } = self;
        format!("Do not use `{origin}.ByteString`, which has unclear semantics and is deprecated")
    }
}

#[derive(Debug, Clone, Copy, Eq, PartialEq)]
enum ByteStringOrigin {
    Typing,
    CollectionsAbc,
}

impl std::fmt::Display for ByteStringOrigin {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(match self {
            Self::Typing => "typing",
            Self::CollectionsAbc => "collections.abc",
        })
    }
}
```

The diagnostic has the invariant that the error message will always be the same, except for the detail of where `ByteString` is coming from; by using an enum here, we make that clearer for readers of the code by encoding the invariant in the type. Using an enum also means that the diagnostic struct will take up less space in memory -- that's not that important here, as the diagnostic structs aren't performance-critical, but it's always nice to use less memory where possible :-)